pkg/backend/filestate/store.go

```diff
@@ -232,49 +235,63 @@ func (p *projectReferenceStore) ListProjects() ([]tokens.Name, error) {
 func (p *projectReferenceStore) ListReferences() ([]*localBackendReference, error) {
 	// The first level of the bucket is the project name.
 	// The second level of the bucket is the stack name.
-	path := StacksDir
-
-	projects, err := p.ListProjects()
-	if err != nil {
-		return nil, err
-	}
+	ctx := context.TODO()
+	prefix := StacksDir + "/"
+	iter := p.bucket.List(&blob.ListOptions{
+		Prefix: prefix,
+		// Don't set the Delimiter.
+		// This will treat the entire bucket as a flat list,
+		// returning only files under the prefix.
+	})
 
 	var stacks []*localBackendReference
-	for _, projName := range projects {
-		// TODO: Could we improve the efficiency here by firstly making listBucket return an enumerator not
-		// eagerly collecting all keys into a slice, and secondly by getting listBucket to return all
-		// descendent items not just the immediate children. We could then do the necessary splitting by
-		// file paths here to work out project names.
-		projectFiles, err := listBucket(p.bucket, filepath.Join(path, projName.String()))
+	for {
+		file, err := iter.Next(ctx)
 		if err != nil {
-			return nil, fmt.Errorf("error listing stacks: %w", err)
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, fmt.Errorf("list bucket: %w", err)
 		}
 
-		for _, projectFile := range projectFiles {
-			// Can ignore directories at this level
-			if projectFile.IsDir {
-				continue
-			}
+		if file.IsDir {
+			continue
+		}
 
-			objName := objectName(projectFile)
-			// Skip files without valid extensions (e.g., *.bak files).
-			ext := filepath.Ext(objName)
-			// But accept gzip compression
-			if ext == encoding.GZIPExt {
-				objName = strings.TrimSuffix(objName, encoding.GZIPExt)
-				ext = filepath.Ext(objName)
-			}
+		// Key is in the form,
+		//   $StacksDir/$projName/$stackName.json[.gz]
+		// We want to extract projName and stackName from it.
 
-			if _, has := encoding.Marshalers[ext]; !has {
-				continue
-			}
+		parts := strings.Split(strings.TrimPrefix(file.Key, prefix), "/")
+		if len(parts) != 2 {
+			continue // skip paths too shallow or too deep
```

Comment:
Nit: maybe comment something of the nature

```suggestion
			continue // skip paths that do not match the specification
```